src/cache_invalidate.c

```diff
@@ -59,45 +60,44 @@ cache_invalidate_relcache_all(void)
 	ts_bgw_job_cache_invalidate_callback();
 }
 
+static Oid hypertable_proxy_table_oid = InvalidOid;
+static Oid bgw_proxy_table_oid = InvalidOid;
+
+void
+ts_cache_invalidate_set_proxy_tables(Oid hypertable_proxy_oid, Oid bgw_proxy_oid)
+{
+	hypertable_proxy_table_oid = hypertable_proxy_oid;
+	bgw_proxy_table_oid = bgw_proxy_oid;
+}
+
 /*
  * This function is called when any relcache is invalidated.
  * Should route the invalidation to the correct cache.
+ *
+ * NOTE that the callback should not call any functions that could invoke the
+ * relcache or syscache to query information during the invalidation. That
+ * might lead to bad things happening.
  */
 static void
-cache_invalidate_callback(Datum arg, Oid relid)
+cache_invalidate_relcache_callback(Datum arg, Oid relid)
 {
-	static bool in_recursion = false;
-
-	if (ts_extension_invalidate(relid))
+	if (!OidIsValid(relid))
 	{
 		cache_invalidate_relcache_all();
```

Comment:
`cache_invalidate_relcache_all()` calls:
`ts_hypertable_cache_invalidate_callback()`->`ts_cache_invalidate(hypertable_cache_current)`->`cache_destroy(cache)`->`cache->pre_destroy_hook(cache)` which calls:
```
static void
connection_cache_entry_free(void *gen_entry)
{
	ConnectionCacheEntry *entry = gen_entry;

	if (entry->conn != NULL)
	{
		/* Cannot directly read the Log_connections boolean since it is not
		 * exported on Windows */
		const char *log_conns = GetConfigOption("log_connections", true, false);

		if (NULL != log_conns && strcmp(log_conns, "on") == 0)
		{
			/* Log the connection closing if requested. Only log the
			 * associated user ID since cannot lookup the user name here due
			 * to, potentially, not being in a proper transaction state where
			 * it is possible to use the syscache. */
			elog(LOG,
				 "closing cached connection to \"%s\" [UserId: %d]",
				 remote_connection_node_name(entry->conn),
				 entry->id.user_id);
		}

		remote_connection_close(entry->conn);
		entry->conn = NULL;
	}
}
```
which recurses into the syscache again.

Reply:
@mfundul Where in this code is the syscache invoked? Is it `GetConfigOption`?

Reply:
Yes, for instance `GetConfigOption()`->`roles_is_member_of()`:
```
	if (!OidIsValid(MyDatabaseId))
		dba = InvalidOid;
	else
	{
		HeapTuple	dbtup;

		dbtup = SearchSysCache1(DATABASEOID, ObjectIdGetDatum(MyDatabaseId));
		if (!HeapTupleIsValid(dbtup))
			elog(ERROR, "cache lookup failed for database %u", MyDatabaseId);
		dba = ((Form_pg_database) GETSTRUCT(dbtup))->datdba;
		ReleaseSysCache(dbtup);
	}
```

Reply:
@mfundul  So, that's syscache, not the relcache. Not sure it makes a difference though. I'll see if I can work around it, but it might require bigger refactoring

Reply:
@mfundul Actually, I fail to see how this can happen. There is no pre_destroy_hook for the hypertable cache. I think you are mixing up the hypertable cache and the connection cache.